File: FactoryMethod_06032026/JavaScript/Ex1.js
// PALETA DE CORES
const ui = {
    reset: "\x1b[0m",
    bold: "\x1b[1m",
    pink: "\x1b[35m",
    purple: "\x1b[38;5;54m",  
    white: "\x1b[97m"
};


// Classe Base de Pessoa
class Pessoa {
    constructor(nome, cargo, permissao) {
        this.nome = nome;
        this.cargo = cargo;
        this.permissao = permissao;
    }

    mostrarDetalhes() {
        console.log(
            ui.pink + "│ " +
            ui.white + this.nome.padEnd(10) +
            ui.pink + " │ " +
            ui.white + this.cargo.padEnd(14) +
            ui.pink + " │ " +
            ui.white + this.permissao.padEnd(10) +
            ui.pink + " │" +
            ui.reset
        );
    }
}

// Subclasses de Pessoa
class Aluno extends Pessoa {
    constructor(nome) {
        super(nome, "Aluno", "Nível 0");
    }
}

class Administrador extends Pessoa {
    constructor(nome) {
        super(nome, "Administrador", "Nível 1");
    }
}

class Professor extends Pessoa {
    constructor(nome) {
        super(nome, "Professor", "Nível 2");
    }
}

class Visitante extends Pessoa {
    constructor(nome) {
        super(nome, "Visitante", "Nível 3");
    }
}

// Criação Abstrata de Pessoa
class CriacaoDePessoas {
    criarPessoa(nome) {
        throw new Error("criarPessoa() deve ser implementado");
    }
}

// Criação concreta de Alunos
class CriacaoDeAlunos extends CriacaoDePessoas {
    criarPessoa(nome) {
        return new Aluno(nome);
    }
}

// Criação concreta de Administradores
class CriacaoDeAdministradores extends CriacaoDePessoas {
    criarPessoa(nome) {
        return new Administrador(nome);
    }
}

// Criação concreta de Professores
class CriacaoDeProfessores extends CriacaoDePessoas {
    criarPessoa(nome) {
        return new Professor(nome);
    }
}

// Criação concreta de Visitantes
class CriacaoDeVisitantes extends CriacaoDePessoas {
    criarPessoa(nome) {
        return new Visitante(nome);
    }
}

// Uso
console.log(ui.pink + ui.bold + "\n════════════ SISTEMA DE PESSOAS ════════════" + ui.reset);

console.log(ui.pink + "┌────────────┬────────────────┬────────────┐" + ui.reset);
console.log(
    ui.pink + "│ " +
    ui.purple + "Nome".padEnd(10) +
    ui.pink + " │ " +
    ui.purple + "Cargo".padEnd(14) +
    ui.pink + " │ " +
    ui.purple + "Permissão".padEnd(10) +
    ui.pink + " │" +
    ui.reset
);
console.log(ui.pink + "├────────────┼────────────────┼────────────┤" + ui.reset);

const criacaoAlunos = new CriacaoDeAlunos();
const criacaoAdministradores = new CriacaoDeAdministradores();
const criacaoProfessores = new CriacaoDeProfessores();
const criacaoVisitantes = new CriacaoDeVisitantes();

const p1 = criacaoAlunos.criarPessoa("Carlos");
const p2 = criacaoProfessores.criarPessoa("Ana");
const p3 = criacaoAdministradores.criarPessoa("João");
const p4 = criacaoVisitantes.criarPessoa("Maria");

p1.mostrarDetalhes();
p2.mostrarDetalhes();
p3.mostrarDetalhes();
p4.mostrarDetalhes();

console.log(ui.pink + "└────────────┴────────────────┴────────────┘" + ui.reset);

File: Criacionais/FactoryMethod/Python/Ex1.py
# PALETA DE CORES
ui = {
    "reset": "\033[0m",
    "bold": "\033[1m",
    "pink": "\033[35m",
    "purple": "\033[38;5;54m",
    "white": "\033[97m"
}

# Classe Base de Pessoa
class Pessoa:
    def __init__(self, nome, cargo, permissao):
        self.nome = nome
        self.cargo = cargo
        self.permissao = permissao

    def mostrar_detalhes(self):
        print(
            f"{ui['pink']}│ {ui['white']}{self.nome.ljust(10)}"
            f"{ui['pink']} │ {ui['white']}{self.cargo.ljust(14)}"
            f"{ui['pink']} │ {ui['white']}{self.permissao.ljust(10)}"
            f"{ui['pink']} │{ui['reset']}"
        )

# Subclasses de Pessoa
class Aluno(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, "Aluno", "Nível 0")

class Administrador(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, "Administrador", "Nível 1")

class Professor(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, "Professor", "Nível 2")

class Visitante(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, "Visitante", "Nível 3")

# Criação Abstrata de Pessoa
class CriacaoDePessoas:
    def criar_pessoa(self, nome):
        raise NotImplementedError("criar_pessoa() deve ser implementado")

# Criações concretas
class CriacaoDeAlunos(CriacaoDePessoas):
    def criar_pessoa(self, nome):
        return Aluno(nome)

class CriacaoDeAdministradores(CriacaoDePessoas):
    def criar_pessoa(self, nome):
        return Administrador(nome)

class CriacaoDeProfessores(CriacaoDePessoas):
    def criar_pessoa(self, nome):
        return Professor(nome)

class CriacaoDeVisitantes(CriacaoDePessoas):
    def criar_pessoa(self, nome):
        return Visitante(nome)

# Uso
print(ui['pink'] + ui['bold'] + "\n════════════ SISTEMA DE PESSOAS ════════════" + ui['reset'])
print(ui['pink'] + "┌────────────┬────────────────┬────────────┐" + ui['reset'])
print(
    ui['pink'] + "│ " +
    ui['purple'] + "Nome".ljust(10) +
    ui['pink'] + " │ " +
    ui['purple'] + "Cargo".ljust(14) +
    ui['pink'] + " │ " +
    ui['purple'] + "Permissão".ljust(10) +
    ui['pink'] + " │" +
    ui['reset']
)
print(ui['pink'] + "├────────────┼────────────────┼────────────┤" + ui['reset'])

criacao_alunos = CriacaoDeAlunos()
criacao_administradores = CriacaoDeAdministradores()
criacao_professores = CriacaoDeProfessores()
criacao_visitantes = CriacaoDeVisitantes()

p1 = criacao_alunos.criar_pessoa("Carlos")
p2 = criacao_professores.criar_pessoa("Ana")
p3 = criacao_administradores.criar_pessoa("João")
p4 = criacao_visitantes.criar_pessoa("Maria")

p1.mostrar_detalhes()
p2.mostrar_detalhes()
p3.mostrar_detalhes()
p4.mostrar_detalhes()

print(ui['pink'] + "└────────────┴────────────────┴────────────┘" + ui['reset'])